src/pcap/api.go

```diff
@@ -0,0 +1,650 @@
+	case codes.InvalidArgument:
+		return newMessageResponse(MessageType_INVALID_REQUEST, err.Error(), target.Identifier)
+	case codes.Aborted:
+		return newMessageResponse(MessageType_INSTANCE_UNAVAILABLE, err.Error(), target.Identifier)
+	case codes.Internal, codes.Unknown:
+		return newMessageResponse(MessageType_CONNECTION_ERROR, err.Error(), target.Identifier)
+	case codes.FailedPrecondition:
+		return newMessageResponse(MessageType_START_CAPTURE_FAILED, err.Error(), target.Identifier)
+	case codes.ResourceExhausted:
+		return newMessageResponse(MessageType_LIMIT_REACHED, err.Error(), target.Identifier)
+	case codes.Unavailable:
+		return newMessageResponse(MessageType_INSTANCE_UNAVAILABLE, err.Error(), target.Identifier)
+	default:
+		return newMessageResponse(MessageType_CONNECTION_ERROR, err.Error(), target.Identifier)
+	}
+}
+
+// boshRequestReceiver is an interface used by boshStopCmd to simplify testing.
+type requestReceiver interface {
+	Recv() (*CaptureRequest, error)
+}
+
+// stopCmd reads the next message from the stream. It ensures that the message
+// has a payload of StopBoshCapture. If any error is encountered or the payload is
+// of a different type an appropriate cause is set and the cancel function is called.
+func stopCmd(cancel CancelCauseFunc, stream requestReceiver) {
+	go func() {
+		msg, err := stream.Recv()
+		if err != nil {
+			cancel(errorf(codes.Unknown, "read message: %w", err))
+			return
+		}
+
+		if msg.GetOperation() == nil {
+			cancel(errorf(codes.InvalidArgument, "read operation: operation was nil: %w", errNilField))
+			return
+		}
+
+		// Gets a Stop message if it's there. Returns nil in any other case, also other message types.
+		stop := msg.GetStop()
+		if stop == nil {
+			cancel(errorf(codes.InvalidArgument, "read operation: expected message of type Stop: %w", errInvalidPayload))
+			return
+		}
+
+		// cancel without cause - normal exit
+		zap.L().Debug("client requested stop of capture")
+		cancel(nil)
+	}()
+}
+
+type streamPreparer interface {
+	prepareStreamToTarget(context.Context, *CaptureOptions, AgentEndpoint, credentials.TransportCredentials, *zap.Logger) (captureReceiver, error)
+}
+
+func (api *API) capture(ctx context.Context, stream responseSender, streamPrep streamPreparer, opts *CaptureOptions, targets []AgentEndpoint, creds credentials.TransportCredentials, log *zap.Logger) (<-chan *CaptureResponse, error) {
+	var captureCs []<-chan *CaptureResponse
+
+	runningCaptures := 0
+	for _, target := range targets {
+		log = log.With(zap.String("target", target.String()))
+		log.Info("starting capture")
+
+		captureStream, err := streamPrep.prepareStreamToTarget(ctx, opts, target, creds, log)
+		if err != nil {
+			errMsg := convertStatusCodeToMsg(err, target)
+			sendErr := stream.Send(errMsg)
+			if sendErr != nil {
+				// FIXME the return interrupts the for-loop over targets and returns in case of send error
```

Comment:
I don't understand what is to be fixed but it should be fixed (if it isn't already)

Reply:
Done